src/Summoner/Source.hs

```diff
@@ -0,0 +1,55 @@
+{-# LANGUAGE ViewPatterns #-}
+
+module Summoner.Source
+       ( Source (..)
+       , sourceT
+       , fetchSource
+       ) where
+
+import Relude
+
+import Control.Arrow ((>>>))
+import Control.Exception (catch)
+import System.Process (readProcess)
+import Toml (BiMap, BiToml, Key)
+
+import Summoner.Ansi (errorMessage)
+
+import qualified Toml
+
+data Source
+    -- | URL link to the source file.
+    = Url Text
+    -- | File path to the local source file.
+    | File FilePath
+    deriving (Show, Eq)
+
+sourceT :: Key -> BiToml Source
+sourceT nm = Toml.match (Toml._Text   >>> _Url)  (nm <> "url")
+         <|> Toml.match (Toml._String >>> _File) (nm <> "file")
+  where
+    _Url :: BiMap Text Source
+    _Url = Toml.invert $ Toml.prism (source Just (const Nothing)) Url
```

Comment:
We need to add some utility function to `tomland` which is like `Toml.invert . Toml.prism` but I don't know good name for such function... 😞 Looks like `prism` function is not needed in most cases and often we need `Toml.invert . Toml.prism`.